src/core/writer.js

```diff
@@ -25,36 +25,68 @@ import { SimpleDOMNode, SimpleXMLParser } from "./xml_parser.js";
 import { BaseStream } from "./base_stream.js";
 import { calculateMD5 } from "./crypto.js";
 
-function writeObject(ref, obj, buffer, transform) {
+async function writeObject(ref, obj, buffer, transform) {
   buffer.push(`${ref.num} ${ref.gen} obj\n`);
   if (obj instanceof Dict) {
-    writeDict(obj, buffer, transform);
+    await writeDict(obj, buffer, transform);
   } else if (obj instanceof BaseStream) {
-    writeStream(obj, buffer, transform);
+    await writeStream(obj, buffer, transform);
   }
   buffer.push("\nendobj\n");
 }
 
-function writeDict(dict, buffer, transform) {
+async function writeDict(dict, buffer, transform) {
   buffer.push("<<");
   for (const key of dict.getKeys()) {
     buffer.push(` /${escapePDFName(key)} `);
-    writeValue(dict.getRaw(key), buffer, transform);
+    await writeValue(dict.getRaw(key), buffer, transform);
   }
   buffer.push(">>");
 }
 
-function writeStream(stream, buffer, transform) {
+async function writeStream(stream, buffer, transform) {
   let string = stream.getString();
   if (transform !== null) {
     string = transform.encryptString(string);
   }
+
+  const filter = stream.dict.getRaw("Filter");
+  const flateDecode = Name.get("FlateDecode");
+  const MIN_LENGTH_FOR_COMPRESSING = 192;
+  let mustCompress = string.length >= MIN_LENGTH_FOR_COMPRESSING;
+
+  if (Array.isArray(filter)) {
+    if (filter.includes(flateDecode)) {
+      mustCompress = true;
+    } else if (mustCompress) {
+      filter.push(flateDecode);
+    }
+  } else if (mustCompress) {
+    if (!filter) {
+      stream.dict.set("Filter", flateDecode);
+    } else if (filter !== flateDecode) {
```

Comment:
There's a subtle problem with this check, since it doesn't actually take the `Name`-caching (in `src/core/primitives.js)` fully into account. Hence the following is probably necessary to be guaranteed to avoid any problems.
```suggestion
    } else if (!(filter instanceof Name) || filter.name !== flateDecode.name) {
```